spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceCDFSuite.scala

```diff
@@ -0,0 +1,460 @@
+      "startingVersion" -> "latest"
+    ))
+    val latestOffset = source.latestOffset(null, source.getDefaultReadLimit)
+    assert(latestOffset == null)
+  }
+
+  integrationTest("DeltaSharingSource.latestOffset - no startingVersion") {
+    val source = getSource(Map("ignoreChanges" -> "true", "ignoreDeletes" -> "true"))
+    val latestOffset = source.latestOffset(null, source.getDefaultReadLimit)
+
+    assert(latestOffset.isInstanceOf[DeltaSharingSourceOffset])
+    val offset = latestOffset.asInstanceOf[DeltaSharingSourceOffset]
+    assert(offset.sourceVersion == 1)
+    assert(offset.tableId == deltaLog.snapshot().metadata.id)
+    assert(offset.tableVersion == 5)
+    assert(offset.index == -1)
+    assert(!offset.isStartingVersion)
+  }
+
+  /**
+   * Test getBatch for Stream CDF
+   */
+  integrationTest("DeltaSharingSource.getBatch - exception") {
+    // getBatch cannot be called without writestream
+    val source = getSource(Map(
+      "ignoreChanges" -> "true",
+      "ignoreDeletes" -> "true",
+      "startingVersion" -> "0"))
+    val latestOffset = source.latestOffset(null, source.getDefaultReadLimit)
+    intercept[AnalysisException] {
+      // Error message would be:
+      //    Queries with streaming sources must be executed with writeStream.start()
+      source.getBatch(None, latestOffset.asInstanceOf[DeltaSharingSourceOffset]).show()
+    }
+  }
+
+  /**
+   * Test basic cdf streaming functionality
+   */
+  integrationTest("CDF Stream basic - success") {
+    var query = withStreamReaderAtVersion()
+      .load().writeStream.format("console").start()
+
+    try {
+      query.processAllAvailable()
+      val progress = query.recentProgress.filter(_.numInputRows != 0)
+      assert(progress.length === 1)
+      progress.foreach { p =>
+        assert(p.numInputRows === 17)
+      }
+    } finally {
+      query.stop()
+    }
+
+    // Test readchangeData=true, and startingVersion=1.
+    query = spark.readStream.format("deltaSharing")
+      .option("readchangeData", "true")
+      .option("startingVersion", "1")
+      .load(cdfTablePath).writeStream.format("console").start()
+    try {
+      query.processAllAvailable()
+      val progress = query.recentProgress.filter(_.numInputRows != 0)
+      assert(progress.length === 1)
+      progress.foreach { p =>
+        assert(p.numInputRows === 17)
+      }
+    } finally {
+      query.stop()
+    }
+  }
+
+  integrationTest("CDF Stream select columns - success") {
+    // select all columns in a different order
+    var query = withStreamReaderAtVersion()
+      .load()
+      .select("birthday", "_commit_version", "name", "_commit_timestamp", "age", "_change_type")
+      .writeStream.format("console").start()
+
+    try {
+      query.processAllAvailable()
+      val progress = query.recentProgress.filter(_.numInputRows != 0)
+      assert(progress.length === 1)
+      progress.foreach { p =>
+        assert(p.numInputRows === 17)
+      }
```

Comment:
can we assert on the number/name of columns?

Reply:
good question. short answer no. 
I added a test to check the schema. 
And as the `format("console")` is used, I eye-ball the output dataframe in the terminal.
There's a TODO to support tmp dir and use it to store streaming checkpoing and stream output, to check dataframe.